Bump golang to 1.11.0

kolyshkin opened the pull request:
It's that time of year again! Go 1.11 is released, let's use it.

This PR also:
* removes our archive/tar fork (since upstream archive/tar is fixed for static builds, and osusergo build tag is set in PR #37500);
* removes `ENV GO_VERSION` from Dockerfile (not needed since PR #37592);
* reformats sources with `gofmt -s` from go 1.11 to avoid validation errors;
* Revendor ugorji/go to include a backported fix (https://github.com/kolyshkin/ugorji-go/commit/1cf431c13dec46596)

![Gopher](https://www.maxpixel.net/static/photo/2x/Animal-Mammal-Nora-Nature-Gopher-Meadow-Spring-2196633.jpg)
_image credit: https://www.maxpixel.net/Animal-Mammal-Nora-Nature-Gopher-Meadow-Spring-2196633_

kolyshkin committed:
Bump golang to 1.11.0

It's that time of year again! Go 1.11 is released, time to use it.

This commit also

* removes our archive/tar fork, since upstream archive/tar
  is fixed for static builds, and osusergo build tag is set.

* removes ENV GO_VERSION from Dockerfile as it's not needed
  anymore since PR #37592 is merged.

[v2: switch to beta2]
[v3: switch to beta3]
[v4: rc1]
[v5: remove ENV GO_VERSION as PR #37592 is now merged]
[v6: rc2]
[v7: final!]
[v8: use 1.11.0]
[v9: back to 1.11]
[v8: use 1.11.0]

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

cpuguy83 commented:
> golang:1.11beta1 not found

kolyshkin commented:
It worked on my laptop before I pushed it. Perhaps CI uses some old(er) private copy of dockerhub that needs to be updated?

thaJeztah commented:
/cc @cyli for the certificate-transparency issue (FYI)

kolyshkin commented:
certificate-transparency requires a one-line patch but it will break 1.10 compatibility. I think a backward-compatible fix can be made, too, although it'd be more than one liner.

thaJeztah commented:
I think it's important we report it (both in the Go issue tracker, and in certificate-transparency); don't think there's been a single version of go that didn't break _something_

kolyshkin commented:
> I think it's important we report it

https://github.com/google/certificate-transparency-go/issues/284
https://github.com/google/certificate-transparency-go/pull/285 (looks somewhat ugly to me)

Not quite sure yet if I want to report this one to golang/go (as the issue they solved is quite complicated) but I left a comment in https://github.com/golang/go/issues/21376#issuecomment-402445810

thaJeztah commented:
Thanks @kolyshkin those look good; at least we reported upstream (so that they're aware of the issue), and we have a possible way forward 👍

thaJeztah commented:
Opened https://github.com/moby/moby/pull/37390 to bump certificate-transparency to match what's currently in SwarmKit (so that a later bump will be smaller)

kolyshkin commented:
Not sure where this is coming from

> 13:27:55 --- FAIL: TestSessionCreate (0.01s)
> 13:27:55     session_test.go:23: assertion failed: error is not nil: Post http://%2Fgo%2Fsrc%2Fgithub.com%2Fdocker%2Fdocker%2Fbundles%2Ftest-integration%2Fdocker.sock/session: net/http: HTTP/1.x transport connection broken: malformed HTTP status code "*"

kolyshkin committed:
Format code with gofmt -s from go-1.11beta1

This should eliminate a bunch of new (go-1.11 related) validation
errors telling that the code is not formatted with `gofmt -s`.

No functional change, just whitespace (i.e.
`git show --ignore-space-change` shows nothing).

Patch generated with:

> git ls-files | grep -v ^vendor/ | grep .go$ | xargs gofmt -s -w

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

thaJeztah commented:
ping @tonistiigi could you have a look at that session failure? IIRC, that endpoint does an upgrade?

kolyshkin commented:
janky fail:

> 11:29:17 WARNING: failed to resolve symlinks in copy.go: lstat /go/src/github.com/docker/docker/copy.go: no such file or directory
> 11:29:17 WARNING: failed to make copy.go a relative path: Rel: can't make copy.go relative to /go/src/github.com/docker/docker
> 11:29:17 WARNING: failed to resolve symlinks in copy.go: lstat /go/src/github.com/docker/docker/copy.go: no such file or directory
> 11:29:17 WARNING: failed to make copy.go a relative path: Rel: can't make copy.go relative to /go/src/github.com/docker/docker
> 11:29:17 WARNING: failed to resolve symlinks in copy.go: lstat /go/src/github.com/docker/docker/copy.go: no such file or directory
> 11:29:17 WARNING: failed to make copy.go a relative path: Rel: can't make copy.go relative to /go/src/github.com/docker/docker
> 11:29:17 WARNING: failed to resolve symlinks in copy.go: lstat /go/src/github.com/docker/docker/copy.go: no such file or directory
> 11:29:17 WARNING: failed to make copy.go a relative path: Rel: can't make copy.go relative to /go/src/github.com/docker/docker
> 11:29:17 copy.go:241::warning: unnecessary conversion (unconvert)
> 11:29:17 copy.go:240::warning: unnecessary conversion (unconvert)
> 11:29:17 copy.go:240::warning: unnecessary conversion (unconvert)
> 11:29:17 copy.go:241::warning: unnecessary conversion (unconvert)
> 11:29:17 Build step 'Execute shell' marked build as failure

This is something going wrong with gometalinter. The file is `daemon/graphdriver/copy/copy.go` but somehow gometalinter (or maybe one of linters) fails to figure out its path, and probably because of this it fails to ignore the ` // nolint: unconvert` directive added by commit 21b2c278cc.

@dnephin @alecthomas ^^^ can you PTAL?

thaJeztah commented:
Should we try updating gometalinter? Looks like the version hasn't been updated in a while; 

https://github.com/moby/moby/blob/49ad3d69a7780e54ace50290d0a1a55c1182a947/hack/dockerfile/install/gometalinter.installer#L3

https://github.com/alecthomas/gometalinter/compare/bfcc1d6942136fd86eb6f1a6fb328de8398fbd80...master (version v2.0.6 was tagged a few days ago https://github.com/alecthomas/gometalinter/releases/tag/v2.0.6)

dnephin commented:
Doesn't look like it helped. I'll try to look into it further soon. I took a quick look and my best guess so far is that something started returning basename instead of a full path to a file. I'm not sure why that is. Maybe a regression in the go beta release?

To get the rest of Ci running maybe add an `|| true` to the validate step for now?

kolyshkin committed:
Revendor ugorji/go

To include the following backported fix:

https://github.com/kolyshkin/ugorji-go/commit/1cf431c13dec46596

which should fix this:

> 13:40:53 vendor/github.com/ugorji/go/codec/gen-helper.generated.go:1:
> possible malformed +build comment%!(EXTRA []interface {}=[])

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

thaJeztah commented:
@dnephin looks like there's more breaking changes in Go 1.11; I opened https://github.com/alecthomas/gometalinter/pull/499 to test gometalinter on 1.11, and https://github.com/golang/go/issues/26222 upstream (output being printed twice)

thaJeztah commented on api/types/strslice/strslice_test.go:
```diff
@@ -29,8 +29,8 @@ func TestStrSliceMarshalJSON(t *testing.T) {
 	parts := map[string][]string{
-		"":   {"default", "values"},
-		"[]": {},
+		"":                        {"default", "values"},
```
As a workaround, we can add a blank line between the short and long keys to make it work for both 1.10 and 1.11 (see https://github.com/containerd/containerd/pull/2436) but it's really a workaround

I opened a ticket upstream for the `gofmt` change; https://github.com/golang/go/issues/26228

thaJeztah replied:
Ok, this was marked as a "won't fix", "documentation only" change https://github.com/golang/go/issues/26228#issuecomment-402684430, so if we want to support multiple versions of Go, we need to make some changes to CI and/or the contributing guide;

> This change was done on purpose in https://github.com/golang/go/commit/542ea5ad91367d2b40589942cc5757a4d5f43f97.
> As to why gofmt changes every release - see @griesemer's comment in https://github.com/golang/go/issues/25161#issuecomment-385481599.
> However, I agree that this change should be documented in the release notes. I can't see it in https://tip.golang.org/doc/go1.11.

kolyshkin replied:
Yes, this looks better (in most cases), thanks for suggestion!

thaJeztah replied:
Discussing possible enhancements on the ticket upstream to ease migrating between Go versions.

kolyshkin replied:
In most cases, if I remember correctly, once rebasing to a newer Go version we are losing compatibility with older ones either directly (i.e. some code in moby/moby starts using some feature from the new Go version) or indirectly (i.e. some code in vendored or otherwise used packages relies on a feature from the new Go version). I am about 84% sure current moby can't be compiled with Go 1.9, for example.

Therefore, chances are slim we'll be supporting Go 1.10 when Go 1.11 will become the current stable version of Go. Having said that, I am all for maintaining compatibility with older versions unless it requires some tremendous efforts.

kolyshkin commented:
Finally something new for a change! On janky, powerpc, and z:

> 12:05:57 --- FAIL: TestMountInit (0.00s)
> 12:05:57     mount_test.go:40: operation not supported
> 12:05:57 --- FAIL: TestMountChanges (0.00s)
> 12:05:57     mount_test.go:152: operation not supported
> 12:05:57 FAIL

On Windows:
> 12:09:14 --- FAIL: TestTarSums (0.05s)
> 12:09:14     tarsum_test.go:393: expecting [tarsum+sha256:8bf12d7e67c51ee2e8306cba569398b1b9f419969521a12ffb9d8875e8836738], but got [tarsum+sha256:75258b2c5dcd9adfe24ce71eeca5fc5019c7e669912f15703ede92b1a60cb11f]
> 12:09:14     tarsum_test.go:393: expecting [tarsum+md5:0d7529ec7a8360155b48134b8e599f53], but got [tarsum+md5:3a6cdb475d90459ac0d3280703d17be2]
> 12:09:14     tarsum_test.go:393: expecting [tarsum+sha1:f1fee39c5925807ff75ef1925e7a23be444ba4df], but got [tarsum+sha1:14b5e0d12a0c50a4281e86e92153fa06d55d00c6]
> 12:09:14     tarsum_test.go:393: expecting [tarsum+sha224:6319390c0b061d639085d8748b14cd55f697cf9313805218b21cf61c], but got [tarsum+sha224:dd8925b7a4c71b13f3a68a0f9428a757c76b93752c398f272a9062d5]
> 12:09:14     tarsum_test.go:393: expecting [tarsum+sha384:a578ce3ce29a2ae03b8ed7c26f47d0f75b4fc849557c62454be4b5ffd66ba021e713b48ce71e947b43aab57afd5a7636], but got [tarsum+sha384:e39e82f40005134bed13fb632d1a5f2aa4675c9ddb4a136fbcec202797e68d2f635e1200dee2e3a8d7f69d54d3f2fd27]
> 12:09:14     tarsum_test.go:393: expecting [tarsum+sha512:e9bfb90ca5a4dfc93c46ee061a5cf9837de6d2fdf82544d6460d3147290aecfabf7b5e415b9b6e72db9b8941f149d5d69fb17a394cbfaf2eac523bd9eae21855], but got [tarsum+sha512:7c56de40b2d1ed3863ff25d83b59cdc8f53e67d1c01c3ee8f201f8e4dec3107da976d0c0ec9109c962a152b32699fe329b2dab13966020e400c32878a0761a7e]
> 12:09:14 FAIL

kolyshkin commented:
Rebased; dropped "vendor: switch to certificate-transparency-go, bump cfssl" commit in favor of #37390.

kolyshkin commented:
> 12:05:57 --- FAIL: TestMountInit (0.00s)
> 12:05:57 mount_test.go:40: operation not supported

The error comes from https://github.com/moby/moby/blob/56b14b8c259640e570becfaffa7aeb3209239991/layer/layer_test.go#L149

and the problem is there's no `lchmod` on Linux. Symlinks have no permissions of its own, nor they can be changed. A quote from symlinks(7) man page:
> On  Linux,  the  permissions  of a symbolic link are not used in any operations; the permissions are always 0777 (read, write, and execute for all user categories), and can't be changed.

Now, the above code calls

https://github.com/containerd/continuity/blob/246e49050efdf45e8f17fbbcf1547ee376f9939e/driver/driver_unix.go#L29-L39

which calls

https://github.com/containerd/continuity/blob/246e49050efdf45e8f17fbbcf1547ee376f9939e/sysx/chmod_linux.go#L10-L12

which calls golang's syscall.Fchmodat(), which was modified in Go 1.11beta to make sure to return the error in case flags are set (repeating the glibc behavior):
https://go-review.googlesource.com/c/go/+/118658

Issue to continuity filed.

kolyshkin commented:
Have `FAIL: TestTarSums` on all CI, not just WIndows as reported earlier.

kolyshkin commented:
> Rebased; dropped "vendor: switch to certificate-transparency-go, bump cfssl" commit in favor of #37390.

Still need to re-vendor certificate-transparency-go to include fix for Windows (https://github.com/google/certificate-transparency-go/issues/284). Done

thaJeztah commented:
> Still need to re-vendor certificate-transparency-go to include fix for Windows

I have two PR's pending in swarmkit to bump as well;

- https://github.com/docker/swarmkit/pull/2688 Update google/certificate-transparency-go to v1.0.10
  - just to align to a tagged release, instead of a commit somewhere in between
- https://github.com/docker/swarmkit/pull/2689 Update google/certificate-transparency-go to v1.0.20

kolyshkin commented:
> Have FAIL: TestTarSums on all CI, not just WIndows as reported earlier.

Caused by https://go-review.googlesource.com/c/go/+/85656, fix to follow.

thaJeztah commented:
@kolyshkin could move these commits to a separate PR? They should be easy to review/merge, are already valid on Go 1.10.x, and keep the diff small;

- https://github.com/moby/moby/pull/37358/commits/1f9d67a3ad78732d3664853d237e68ce4d2a432a "daemon/*.go: fix some Wrap[f]/Warn[f] errors"
- https://github.com/moby/moby/pull/37358/commits/1f9d67a3ad78732d3664853d237e68ce4d2a432a "Compile fix"
- https://github.com/moby/moby/pull/37358/commits/6ef7ca23f924b8762cfeebb14a31db5ae55419c5 "aufs: fix Wrapf args"
- https://github.com/moby/moby/pull/37358/commits/c011fc0a800c1273f18a8c9864839100534d7945 "loggerutils: build fixes, improve errors"

thaJeztah commented:
Opened https://github.com/moby/moby/pull/37437 and https://github.com/moby/moby/pull/37436 to separate out those changes

kolyshkin commented:
> Caused by https://go-review.googlesource.com/c/go/+/85656, fix to follow.

This is due to changing `Typeflag` default from `TypeRegA` to `TypeReg`, which results in changed `Typeflag` and also `Checksum` fields in the header.

Now, it is impossible to set `TypeRegA` as Typeflag with Go 1.11 archive/tar, it will be changed to `TypeReg`. So the fix is to set TypeReg explicitly (and change the expected checksums in test cases).

The alternative is to vendor archive/tar again (for the bazillionth time).

kolyshkin commented:
Rebased after merging #37437 (% formatting fixes commits removed); added tarsum fix, reshuffled commits a bit.

kolyshkin commented:
Added a fix for lcow logging (Windows); will separate out to a PR once tested.

thaJeztah commented on pkg/tarsum/tarsum_test.go:
```diff
@@ -84,27 +84,27 @@ var testLayers = []testLayer{
 	{
 		options: &sizedOptions{1, 1024 * 1024, false, false}, // a 1mb file (in memory)
-		tarsum:  "tarsum+sha384:a578ce3ce29a2ae03b8ed7c26f47d0f75b4fc849557c62454be4b5ffd66ba021e713b48ce71e947b43aab57afd5a7636",
+		tarsum:  "tarsum+sha384:e39e82f40005134bed13fb632d1a5f2aa4675c9ddb4a136fbcec202797e68d2f635e1200dee2e3a8d7f69d54d3f2fd27",
```
ping @tonistiigi @dmcgowan for input/impact

kolyshkin replied:
It seems that `FileInfoHeader()` from `pkg/archive` is not affected, as it uses `tar.FileInfoHeader()` which explicitly sets `Typeflag = TypeReg` for regular files.

kolyshkin replied:
In other words, tarballs generated by docker always have `Typeflag = TypeReg` for regular files (quickly checked with `docker save`).

dmcgowan replied:
> for input/impact

thaJeztah replied:
I think the goal previous times was to not break the build cache too many times; given that we'll break the cache when we make the move to buildkit (non experimental), likely we want to avoid breaking the cache before that, but indeed could use Tõnis input on that

kolyshkin replied:
From my understanding (see analisys in comments above), build cache won't be broken -- the checksum issue is limited to this particular test. But this is theory; is there any way to quickly check it?

kolyshkin commented:
Looking at this failure:

> 13:27:55 --- FAIL: TestSessionCreate (0.01s)
> 13:27:55 session_test.go:23: assertion failed: error is not nil: Post http://%2Fgo%2Fsrc%2Fgithub.com%2Fdocker%2Fdocker%2Fbundles%2Ftest-integration%2Fdocker.sock/session: net/http: HTTP/1.x transport connection broken: malformed HTTP status code "*"

Checked docker.log, related messages look similar to ones from the passed build.

Checked net/http differences from go-1.10.2 to 1.11beta1, there are too many and I don't see any clues.

Currently out of ideas.

kolyshkin commented:
Finally got some green here! The powerpc failure

> 21:30:07 FAIL: docker_api_swarm_test.go:296: DockerSwarmSuite.TestAPISwarmLeaderElection

is https://github.com/moby/moby/issues/32673.

Remaining issues are:
* TestSessionCreate failure (https://github.com/moby/moby/pull/37358#issuecomment-402533142)
* gometalinter issue (https://github.com/moby/moby/pull/37358#issuecomment-402547986) _possibly_ caused by https://github.com/golang/go/issues/26125

kolyshkin commented:
Separated continuity bump to a separate PR: https://github.com/moby/moby/pull/37445

kolyshkin commented:
lcow fix separated out to PR https://github.com/moby/moby/pull/37446

kolyshkin commented:
A few more commits separated out:
* pkg/tarsum: fix unit test for Go 1.11+ (https://github.com/moby/moby/pull/37498)
* vendor: bump google/certificate-transparency-go to 1.0.20 (https://github.com/moby/moby/pull/37499)
* Add osusergo build tar for static binaries (https://github.com/moby/moby/pull/37500)

Once merged I will rebase/shrink this one

kolyshkin commented:
CI failure on janky which I haven't seen before:
```
19:29:39 The result of hack/generate-swagger-api.sh differs
19:29:39 
19:29:39 diff --git a/api/types/container/container_wait.go b/api/types/container/container_wait.go
19:29:39 index 9e3910a6b..06b0f0207 100644
19:29:39 --- a/api/types/container/container_wait.go
19:29:39 +++ b/api/types/container/container_wait.go
19:29:39 @@ -7,14 +7,6 @@ package container
19:29:39  // See hack/generate-swagger-api.sh
19:29:39  // ----------------------------------------------------------------------------
19:29:39  
19:29:39 -// ContainerWaitOKBodyError container waiting error, if any
19:29:39 -// swagger:model ContainerWaitOKBodyError
19:29:39 -type ContainerWaitOKBodyError struct {
19:29:39 -
19:29:39 -	// Details of an error
19:29:39 -	Message string `json:"Message,omitempty"`
19:29:39 -}
19:29:39 -
19:29:39  // ContainerWaitOKBody OK response to ContainerWait operation
19:29:39  // swagger:model ContainerWaitOKBody
19:29:39  type ContainerWaitOKBody struct {
19:29:39 @@ -27,3 +19,11 @@ type ContainerWaitOKBody struct {
19:29:39  	// Required: true
19:29:39  	StatusCode int64 `json:"StatusCode"`
19:29:39  }
19:29:39 +
19:29:39 +// ContainerWaitOKBodyError container waiting error, if any
19:29:39 +// swagger:model ContainerWaitOKBodyError
19:29:39 +type ContainerWaitOKBodyError struct {
19:29:39 +
19:29:39 +	// Details of an error
19:29:39 +	Message string `json:"Message,omitempty"`
19:29:39 +}
19:29:39 
19:29:39 Please update api/swagger.yaml with any api changes, then 
19:29:39 run `hack/generate-swagger-api.sh`.
```

I can reproduce it by running swagger about a dozen times -- and once in a dozen it puts ContainerWaitOKBodyError after ContainerWaitOKBody.

The inconsistency is probably caused by https://github.com/go-swagger/go-swagger/issues/846 and is fixed by https://github.com/go-swagger/go-swagger/pull/848 (which is included in version 0.9.0, https://github.com/go-swagger/go-swagger/blob/e341e616eaee0b6bd888b6ba6eb0872d093a23ed/notes/v0.9.0.md).

This indeed fixes the consistency problem, but causes a different one -- an extra set of `Validate()` methods are generated (despite `--skip-validator` flag) which, I guess, are not needed. It is reported in https://github.com/go-swagger/go-swagger/issues/924. Apparently the fix is included in v0.10.0 (https://github.com/go-swagger/go-swagger/blob/7472ff1d09375c648d783a8389c95502fca74378/notes/v0.10.0.md), but it's not working for me.

I have also found out that `--skip-validator` option was silently removed later (in 0.14 if I'm correct, https://github.com/go-swagger/go-swagger/commit/d2aae8150c526d861828ee0e8a16623df01d3fb5#commitcomment-29777263).

I have tried several versions of swagger, and haven't had success with any of them. @dnephin do you have any input on the issue?

kolyshkin commented:
I have already prepared the commit, will push as soon as images will be available on dockerhub :)

kolyshkin commented:
Switched to beta2, rebased, commit `0004-Temporary-ignore-failures-from-gometalinter.patch` temporary removed to see if gometalinter failure was caused by beta1 and is hopefully fixed in beta2.

thaJeztah commented:
```
ERROR: Failed 'ERROR: Mismatched GO versions between Dockerfile and Dockerfile.windows. Update your PR to ensure that both files are updated and in sync. 1.11beta1 1.11beta2

kolyshkin commented:
> ERROR: Failed 'ERROR: Mismatched GO versions between Dockerfile and Dockerfile.windows. Update your PR to ensure that both files are updated and in sync. 1.11beta1 1.11beta2

I hoped to fixed this misfire before anyone notices, and I was pretty sure I fixed it in an instant, yet Seb was faster :rofl:. This strongly enforces my suspicion he has a secret group of people secretly working for him ([like ANK](https://openvz.livejournal.com/48014.html)).

kolyshkin commented:
One gometalinter issue still remains as it was -- a failure to normalize the path of `daemon/graphdriver/copy/copy.go`, and thus a failure to ignore the errors annotated with `nolint:unconvert`. Here is the relevant portion of the file:

https://github.com/moby/moby/blob/1ef1cc8388165b2b848f9b3f53ec91c87de09f63/daemon/graphdriver/copy/copy.go#L236-L241

Here are the error messages from gometalinter:
```
09:45:34 WARNING: failed to make copy.go a relative path: lstat copy.go: no such file or directory
09:45:34 WARNING: failed to make copy.go a relative path: lstat copy.go: no such file or directory
09:45:34 WARNING: failed to make copy.go a relative path: lstat copy.go: no such file or directory
09:45:34 WARNING: failed to make copy.go a relative path: lstat copy.go: no such file or directory
09:46:44 copy.go:241::warning: unnecessary conversion (unconvert)
09:46:44 copy.go:241::warning: unnecessary conversion (unconvert)
09:46:44 copy.go:240::warning: unnecessary conversion (unconvert)
09:46:44 copy.go:240::warning: unnecessary conversion (unconvert)
09:46:45 Build step 'Execute shell' marked build as failure
```

kolyshkin commented:
gometalinter issue 
```
copy.go:240:0: unnecessary conversion
copy.go:240:0: unnecessary conversion
copy.go:241:0: unnecessary conversion
copy.go:241:0: unnecessary conversion
```
sorted out here: https://github.com/mdempsky/unconvert/pull/30

kolyshkin commented:
It seems like all the remaining failures are either flaky tests of some CI glitches.

kolyshkin commented:
Re-push to trigger CI

kolyshkin commented:
Funny, same failure on powerpc and z

...
12:19:36 PASS: docker_api_swarm_test.go:201: DockerSwarmSuite.TestAPISwarmPromoteDemote	15.719s
12:21:01 
12:21:01 ----------------------------------------------------------------------
12:21:01 FAIL: check_test.go:347: DockerSwarmSuite.TearDownTest
12:21:01 
12:21:01 check_test.go:352:
12:21:01     d.Stop(c)
12:21:01 /go/src/github.com/docker/docker/internal/test/daemon/daemon.go:401:
12:21:01     t.Fatalf("Error while stopping the daemon %s : %v", d.id, err)
12:21:01 ... Error: Error while stopping the daemon de1d200a57725 : exit status 130
12:21:01 
12:21:01 
12:21:01 ----------------------------------------------------------------------
12:21:01 PANIC: docker_api_swarm_test.go:359: DockerSwarmSuite.TestAPISwarmRaftQuorum
12:21:01 
12:21:01 [de1d200a57725] waiting for daemon to start
12:21:01 [de1d200a57725] daemon started
12:21:01 
12:21:01 [d8d1db6c7402b] waiting for daemon to start
12:21:01 [d8d1db6c7402b] daemon started
12:21:01 
12:21:01 [dcc72e34f5265] waiting for daemon to start
12:21:01 [dcc72e34f5265] daemon started
12:21:01 
12:21:01 [d8d1db6c7402b] exiting daemon
12:21:01 [dcc72e34f5265] exiting daemon
12:21:01 [d8d1db6c7402b] waiting for daemon to start
12:21:01 [d8d1db6c7402b] daemon started
12:21:01 
12:21:01 [de1d200a57725] daemon started
12:21:01 Attempt #2: daemon is still running with pid 20165
12:21:01 Attempt #3: daemon is still running with pid 20165
12:21:01 Attempt #4: daemon is still running with pid 20165
12:21:01 [de1d200a57725] exiting daemon
12:21:01 ... Panic: Fixture has panicked (see related PANIC)
12:21:01 
12:21:01 ----------------------------------------------------------------------
12:21:01 MISS: docker_api_swarm_test.go:780: DockerSwarmSuite.TestAPISwarmRestartCluster
...

Logs:
* z: https://jenkins.dockerproject.org/job/Docker-PRs-s390x/10497/console
* ppc: https://jenkins.dockerproject.org/job/Docker-PRs-powerpc/10621/console

added a comment to https://github.com/moby/moby/issues/34988#issuecomment-408120424

kolyshkin commented:
The timestamps in the above example look weird as they show the same time. In particular,

> 12:21:01 Attempt #2: daemon is still running with pid 20165
> 12:21:01 Attempt #3: daemon is still running with pid 20165
> 12:21:01 Attempt #4: daemon is still running with pid 20165

look suspicious since there should be 1 second delay between those. Can be a trouble with logging or a manifestation of something bigger going on.

thaJeztah commented:
Looks like there's still discussion on the filename change; 
https://github.com/golang/go/issues/26671#issuecomment-409432591 (I see the issue is reopened)

kolyshkin commented:
Updated to beta3, should be smooth now

> Mysterious DockerSwarmSuite.TearDownTest failure on z/ppc

is gone

> gometalinter uncovert fix from mdempsky/unconvert#31.

The issue is fixed in go-1.11beta3, so the patch is dropped.

kolyshkin commented:
With go-1.11beta3 the only CI failure is on ppc:

> 14:40:15 FAIL: docker_cli_swarm_test.go:1376: DockerSwarmSuite.TestSwarmClusterRotateUnlockKey

This one is flaky: https://github.com/moby/moby/issues/33041#issuecomment-378311901, but it still makes sense to figure out why. Added a commit showing combined output in case the command fails.

kolyshkin commented:
rc1 update went fine, failure in experimental
> 01:10:35.549 FAIL: docker_api_swarm_test.go:296: DockerSwarmSuite.TestAPISwarmLeaderElection
is https://github.com/moby/moby/issues/32673

kolyshkin commented:
> > 14:40:15 FAIL: docker_cli_swarm_test.go:1376: DockerSwarmSuite.TestSwarmClusterRotateUnlockKey
> 
> This one is flaky: https://github.com/moby/moby/issues/33041#issuecomment-378311901, but it still makes sense to figure out why. Added a commit showing combined output in case the command fails.

Separated out to https://github.com/moby/moby/pull/37635

kolyshkin commented:
updated to rc2 -- let's give it a final run before switching to 1.11 final release (which went out already but not yet available from docker hub).

thaJeztah commented:
Looks like there's a genuine failure (failing on all Linux runs);

```
20:17:04 --- FAIL: TestMountInit (0.00s)
20:17:04     mount_test.go:40: operation not supported
20:17:04 --- FAIL: TestMountChanges (0.00s)
20:17:04     mount_test.go:152: operation not supported
20:17:04 FAIL
```

kolyshkin commented:
> 20:17:04 --- FAIL: TestMountInit (0.00s)

Can't repro locally, pushed a debug commit to see what is going on

kolyshkin commented:
> 00:00:41.385 ERROR: Failed 'You cannot call a method on a null-valued expression.' at 08/27/2018 19:37:24

This one's a mystery, I can't even figure out what script is this (does not look like hack/make.ps1).

@johnstep can you PTAL once time permits?

kolyshkin commented:
> 00:10:45.838 --- FAIL: TestMountInit (0.00s)
> 00:10:45.838     mount_test.go:40: lchmod /tmp/graph-434114565/vfs/dir/13916bdc3464afafcddf21a35eaaded550246f61977f49e14e74588bb7e9ffb1-init/testfile.txt: operation not supported
> 00:10:45.838 --- FAIL: TestMountChanges (0.00s)
> 00:10:45.838     mount_test.go:152: lchmod /tmp/graph-230072867/vfs/dir/8f0a0507d36ea9429959bf36b1eb237cc8445bccdfcbd649fd05045f70558042-init/testfile1.txt: operation not supported

Oh. Hmm. Somehow commit 98e23f2a8ef5703861eb98bf78c0354f2ece8c0f rolls back containerd/continuity. Hmm. https://github.com/moby/moby/pull/37664

kolyshkin committed:
vendor: bump containerd/continuity

This is to include the Go 1.11 fix
(https://github.com/containerd/continuity/pull/120).
Again (see c64a2448d1f2b7ba680d).

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
Reshuffled the commits; switched to go 1.11 final!

kolyshkin commented:
> 00:00:36.479 Step 1/93 : FROM golang:1.11 AS base
> 00:00:36.717 1.11: Pulling from library/golang
> 00:00:36.964 no matching manifest for unknown in the manifest list entries

Too early... although it works fine on my laptop (default configuration, i.e. using docker hub).

kolyshkin commented:
Seems the golang docker image is only available for amd64 as of now. Will trigger CI later once other arch images will be available.

kolyshkin commented:
Trying to address Windows CI early failure here (https://github.com/jhowardmsft/docker-w2wCIScripts/pull/29).

kolyshkin commented:
exp and z CI failure looks unrelated (https://github.com/moby/moby/issues/33041):

exp:
> 01:39:45.330 FAIL: docker_cli_swarm_test.go:1376: DockerSwarmSuite.TestSwarmClusterRotateUnlockKey

z:
> 01:54:00.671 FAIL: check_test.go:347: DockerSwarmSuite.TearDownTest
> 01:54:00.672 PANIC: docker_cli_swarm_test.go:1376: DockerSwarmSuite.TestSwarmClusterRotateUnlockKey

thaJeztah commented on Dockerfile.simple:
```diff
@@ -41,8 +41,7 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
-# IMPORTANT: When updating this please note that stdlib archive/tar pkg is vendored
-ENV GO_VERSION 1.10.3
+ENV GO_VERSION 1.11
 RUN curl -fsSL "https://golang.org/dl/go${GO_VERSION}.linux-amd64.tar.gz" \
```
Wondering if we should change the `FROM` to `FROM golang:1.11` or `FROM golang:1.11.0-stretch` instead

jhowardmsft replied:
Just to note, if you change this, I'd have to update the Windows CI script again. It currently copes with the old `ENV GOVERSION=` and `FROM golang:x.xx` only.

kolyshkin replied:
I was thinking about it, too, but it's out of scope for this PR; will do it later.

kolyshkin replied:
In fact the change is pretty small so I've added it to this PR. Thanks!

thaJeztah commented on hack/validate/vendor:
```diff
@@ -10,9 +10,8 @@ validate_vendor_diff(){
-		# for details.
-		ls -d vendor/* | grep -v vendor/archive | xargs rm -rf
+		# cause the validation to fail.
+		ls -d vendor/* | xargs rm -rf
```
If we no longer need to keep the `vendor/archive` dependency; should we bump the version of `vndr` to the latest version, and remove this line? (latest version cleans the vendor dir if you do a full `vendor`)

kolyshkin replied:
Will address as a followup.

kolyshkin replied:
Added

thaJeztah commented on vendor.conf:
```diff
@@ -57,7 +57,7 @@ github.com/samuel/go-zookeeper d0e0d8e11f318e000a8cc434616d69e329edc374
 github.com/coreos/etcd v3.2.1
 github.com/coreos/go-semver v0.2.0
-github.com/ugorji/go f1f1a805ed361a0e078bb537e4ea78cd37dcf065
+github.com/ugorji/go go111 https://github.com/kolyshkin/ugorji-go
```
Can you add a comment referring to the PR we're waiting on to be merged upstream? (Or a comment explaining why we're using a fork here)

kolyshkin replied:
git blame shows it clearly; anyway, added :thumbsup:

thaJeztah commented on vendor/github.com/containerd/continuity/vendor.conf:
```diff
@@ -10,4 +10,4 @@ github.com/spf13/pflag 4c012f6dcd9546820e378d0bdda4d8fc772cdfea
 golang.org/x/net a337091b0525af65de94df2eb7e98bd9962dcbe2
 golang.org/x/sync 450f422ab23cf9881c94e2db30cac0eb1b7cf80c
-golang.org/x/sys 665f6529cca930e27b831a0d1dafffbe1c172924
+golang.org/x/sys 77b0e4315053a57ed2962443614bdb28db152054
```
✅ Checked; master is already ahead of this commit (this is what's on master https://github.com/golang/sys/compare/77b0e4315053a57ed2962443614bdb28db152054...ac767d655b305d4e9612f5f6e33120b9176c4ad4)

thaJeztah reviewed (changes requested):
Thanks @kolyshkin; I left some minor comments

I also opened https://github.com/moby/moby/pull/37727 to bump Go to 1.10.4, which I think would be good to merge first (in case we want to backport that bump to currently supported versions of Docker 😅)

thaJeztah commented:
No idea what's failing on WindowsRS1 btw

jhowardmsft commented:
> No idea what's failing on WindowsRS1 btw

Me neither - running this locally to figure it out.

jhowardmsft commented:
Ah, it's because it's looking for `ENV GOVERSION X.X.X` which has been removed from `Dockerfile`, replaced by `FROM golang:1.11 AS base`.

jhowardmsft commented:
@kolyshkin @thaJeztah Restarted WindowsRS1 CI - verified it got past the previous error now the job is fixed.

kolyshkin committed:
Dockerfile.simple: simplify by using golang img

Instead of installing golang from sources, it's easier to use
golang image which is based on Debian Stretch.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
Bump vndr

We can do that now as we're no longer carrying archive/tar.
Note that latest vndr removes vendor/ subdir so we don't have to,
thus the change in hack/validate/vendor.

While at it, re-run a new vndr version to make sure everything
that should be there is.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
Changed the version from 1.11 to 1.11.0; addressed @thaJeztah comments; rebased.

kolyshkin commented:
OK, it seems I can't use version `1.11.0` where golang is downloaded manually (i.e. in `Dockerfile.windows`). I can't use `1.11.0` in `Dockerfile` as there is a Windows check that the version is the same as Dockerfile.

Rolling back to `1.11` (no `.0`).

jhowardmsft commented:
Yes, it really should be 1.11, not 1.11.0. This is the versioning schema they use.

kolyshkin commented:
ppc CI failure is unrelated (https://github.com/moby/moby/issues/34988):
> 01:59:12.995 FAIL: docker_api_swarm_test.go:359: DockerSwarmSuite.TestAPISwarmRaftQuorum

kolyshkin commented:
> Yes, it really should be 1.11, not 1.11.0. This is the versioning schema they use.

It is different for docker images: for those, 1.11 means "the latest from 1.11.x series" (which currently means 1.11), while 1.11.0 means exactly "1.11". So, to not inadvertently switch to 1.11.1 once it's released, 1.11.0 is required, but it won't work for `Dockerfile.windows`. Could be ironed out by removing `.0` suffix from the `$goVersionDockerfile` but the code there is already complicated... so we're ignoring the issue for now.

kolyshkin commented:
:m: :e-mail: :registered: :gem: :e-mail:

kolyshkin commented:
rebased

kolyshkin commented:
CI on power revealed a (most probably unrelated but) real bug, filed https://github.com/moby/moby/issues/37735

jhowardmsft commented:
Running Windows CI locally also revealed a bug: https://github.com/moby/moby/pull/37770

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
Discussing a bit more in the maintainers meeting; and ideally we'd still have the `.0` in the `FROM` line, but that requires some changes in the `make.ps1`; @kolyshkin will have a look into changing that.

kolyshkin committed:
hack/make.ps1: remove the .0 suffix from go version

We would like to use a version with .0 suffix (like 1.11.0) in
Dockerfile, so that once a .1 version is out (like 1.11.1) we
won't accidentally switch to it.

Unfortunately it's not possible to use .0 suffix currently
as it breaks the check in make.ps1. This patch fixes that.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
To allow for `FROM golang:1.11.0`, we need to remove `.0` suffix from the version number.

To that effect:
* https://github.com/moby/moby/pull/37358/commits/8ad648b59a2993156586458cc9ff8dcbc89b86e4
* https://github.com/jhowardmsft/docker-w2wCIScripts/pull/31

Once the CI PR Is merged, windows CI needs to be restarted, and once green, we're good to go!

jhowardmsft commented:
- [x] CI Script updated merged in my repo
 - [x] Jenkins Windows RS1 job updated with revised script
 - [x] rebuild/WindowsRS1

 - [x] I also restarted experimental failing for an unrelated reason

thaJeztah commented on vendor.conf:
```diff
@@ -57,7 +57,8 @@ github.com/samuel/go-zookeeper d0e0d8e11f318e000a8cc434616d69e329edc374
 github.com/coreos/etcd v3.2.1
 github.com/coreos/go-semver v0.2.0
-github.com/ugorji/go f1f1a805ed361a0e078bb537e4ea78cd37dcf065
+# fix for go vet (https://github.com/kolyshkin/ugorji-go/commit/1cf431c13dec46596)
```
we should create a tracking issue for this (ie to get rid of the fork)

kolyshkin replied:
Sure, https://github.com/moby/moby/issues/37804

thaJeztah reviewed (approved):
LGTM, and all green now!
